### PR TITLE
python jails talk

### PR DESCRIPTION
**Pull request by sevenmaster**

### CODE REVIEW

File: _posts/2023-07-07-python-jails.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: "Python Jail Escapes"
+categories: learning
+author: Liam
+---
+
+At our weekly meetings we had a talk about Python jail escapes, aka. getting around restrictions that make it hard to execute `os.system('cat flag.txt')`. In the talk we went through challenges, that we present here as exercises to practice. Starting very simple and then digging more and more into python internal.
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:8]: internal**s**

File: _posts/2023-07-07-python-jails.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+Solve them yourself, by copying the script and providing input that executes `os.system('cat flag.txt')`. Afterwards you can compare your solution to the one we had in mind, there are always different ways. For some reference solution it is relevant to mention, that we used Python 3.11.2.
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:10]: "For some reference solution**s**" ?

File: _posts/2023-07-07-python-jails.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+Leaning about Python Jail Escapes it is helpful to ask questions like:
+* How are modules and classes implemented in Python?
+* What are frozen modules?
+* How does syntactic sugar get replaced? You may start at `import someting` or `array[5]`.
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:15]: `import something`

File: _posts/2023-07-07-python-jails.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+forbidden = ['eval', 'exec', 'import', 'open', 'system', 'os', 'builtins']
+for f in forbidden:
+    code = code.replace(f, '')
+exec(code)
+```
+
+<details>
+<summary>Reference Solution</summary>
+
+No need for recursivity. This is Python not Haskell.
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:36]: recursiveness

File: _posts/2023-07-07-python-jails.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        print('You are jailed!', keyword)
+        break
+else:
+    exec(code, {'__builtins__': {}})
+```
+
+<details>
+<summary>Reference Solution</summary>
+
+Are there simpler solution to construct a string? Yes. Does this one get you the flag? Also, yes. So why are you asking?
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:105]: solution**s**

File: _posts/2023-07-07-python-jails.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        print('You are jailed!', keyword)
+        break
+else:
+    exec(code)
+```
+
+<details>
+<summary>Reference Solution</summary>
+
+it is not over until you remove getattr
```
Comment (Ik0ri4n) [inline _posts/2023-07-07-python-jails.md:127]: Just style, but every other solution starts with a capital letter

**Review by Ik0ri4n — COMMENTED** [review]
I've commented on a few typos. The post itself seems great to me :smile: 
I didn't try to validate your solutions though, I hope you've already done that

File: _posts/2023-07-07-python-jails.md (hunk 1 above)
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:8]:
```suggestion
At our weekly meetings we had a talk about Python jail escapes, aka. getting around restrictions that make it hard to execute `os.system('cat flag.txt')`. In the talk we went through challenges, that we present here as exercises to practice. Starting very simple and then digging more and more into python internals.
```

File: _posts/2023-07-07-python-jails.md (hunk 2 above)
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:10]:
```suggestion
Solve them yourself, by copying the script and providing input that executes `os.system('cat flag.txt')`. Afterwards you can compare your solution to the one we had in mind, there are always different ways. For some reference solutions it is relevant to mention, that we used Python 3.11.2.
```

File: _posts/2023-07-07-python-jails.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+title: "Python Jail Escapes"
+categories: learning
+author: Liam
+---
+
+At our weekly meetings we had a talk about Python jail escapes, aka. getting around restrictions that make it hard to execute `os.system('cat flag.txt')`. In the talk we went through challenges, that we present here as exercises to practice. Starting very simple and then digging more and more into python internal.
+
+Solve them yourself, by copying the script and providing input that executes `os.system('cat flag.txt')`. Afterwards you can compare your solution to the one we had in mind, there are always different ways. For some reference solution it is relevant to mention, that we used Python 3.11.2.
+
+Leaning about Python Jail Escapes it is helpful to ask questions like:
```
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:12]:
```suggestion
Learning about Python Jail Escapes it is helpful to ask questions like:
```

File: _posts/2023-07-07-python-jails.md (hunk 3 above)
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:15]:
```suggestion
* How does syntactic sugar get replaced? You may start at `import something` or `array[5]`.
```

File: _posts/2023-07-07-python-jails.md (hunk 4 above)
Comment (intrigus-lgtm) [reply _posts/2023-07-07-python-jails.md:36]: It's fine as is, if we want to believe Google Ngram:
https://books.google.com/ngrams/graph?content=recursivity%2Crecursiveness&year_start=1800&year_end=2019&corpus=en-2019&smoothing=3

File: _posts/2023-07-07-python-jails.md (hunk 5 above)
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:105]:
```suggestion
Are there simpler solutions to construct a string? Yes. Does this one get you the flag? Also, yes. So why are you asking?
```

File: _posts/2023-07-07-python-jails.md (hunk 6 above)
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:127]:
```suggestion
It is not over until you remove `getattr`:
```

File: _posts/2023-07-07-python-jails.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+dname = os.path.dirname(abspath)
+os.chdir(dname)
+
+print("-- Please enter code (last line must contain only --END)")
+source_code = ""
+while True:
+  line = sys.stdin.readline()
+  if '(' in line or ')' in line:
+        print("ERROR: Parentheses are not allowed")
+        exit(1)
```
Comment (intrigus-lgtm) [inline _posts/2023-07-07-python-jails.md:216]: Feel very free to just ignore that one suggestion.
```suggestion
  ### Begin: Additional restriction
  if '(' in line or ')' in line:
        print("ERROR: Parentheses are not allowed")
        exit(1)
   ### End: Additional restriction
```

**Review by intrigus-lgtm — APPROVED** [review]
LGTM, modulo some of my and Dominik's comments that I converted to suggestions so you can apply them all in one commit.

File: _posts/2023-07-07-python-jails.md (hunk 4 above)
Comment (Ik0ri4n) [reply _posts/2023-07-07-python-jails.md:36]: Really?! Wow...

**Pull request merged**